migration/src/main/scala/akka/persistence/r2dbc/migration/MigrationTool.scala

```diff
@@ -0,0 +1,324 @@
+          .recoverWith { case _: R2dbcDataIntegrityViolationException =>
+            // events already exists, which is ok, but since the batch
+            // failed we must try again one-by-one
+            Future.sequence(events.map { event =>
+              targetJournalDao.writeEvents(List(event)).recoverWith { case _: R2dbcDataIntegrityViolationException =>
+                // ok, already exists
+                log
+                  .debug("event already exists, persistenceId [{}], seqNr [{}]", event.persistenceId, event.sequenceNr)
+                Future.successful(())
+              }
+            })
+          }
+          .map(_ => events.last.sequenceNr -> events.size)
+      }
+      .mapAsync(1) { case (seqNr, count) =>
+        migrationDao
+          .updateEventProgress(persistenceId, seqNr)
+          .map(_ => count)
+      }
+      .runWith(Sink.fold(0L) { case (acc, count) => acc + count })
+  }
+
+  private def serializedJournalRow(env: EventEnvelope): SerializedJournalRow = {
+    val event = env.event.asInstanceOf[AnyRef]
+    val serialized = serialization.serialize(event).get
+    val serializer = serialization.findSerializerFor(event)
+    val manifest = Serializers.manifestFor(serializer, event)
+
+    val metadata =
+      env.eventMetadata.map { meta =>
+        val m = meta.asInstanceOf[AnyRef]
+        val serializedMeta = serialization.serialize(m).get
+        val metaSerializer = serialization.findSerializerFor(m)
+        val metaManifest = Serializers.manifestFor(metaSerializer, m)
+        SerializedEventMetadata(metaSerializer.identifier, metaManifest, serializedMeta)
+      }
+
+    SerializedJournalRow(
+      env.persistenceId,
+      env.sequenceNr,
+      JournalDao.EmptyDbTimestamp, // FIXME preserve the timestamp, requires some changes to JournalDao
+      JournalDao.EmptyDbTimestamp,
+      serialized,
+      serializer.identifier,
+      manifest,
+      "",
+      metadata)
+  }
+
+  /**
+   * Migrate latest snapshot for a single persistence id.
+   */
+  def migrateSnapshot(persistenceId: String): Future[Int] = {
+    for {
+      _ <- createProgressTable
+      currentProgress <- migrationDao.currentProgress(persistenceId)
+      snapCount <- migrateSnapshot(persistenceId, currentProgress)
+    } yield snapCount
+  }
+
+  private def migrateSnapshot(persistenceId: String, currentProgress: Option[CurrentProgress]): Future[Int] = {
+    val progressSeqNr = currentProgress.map(_.snapshotSeqNr).getOrElse(0L)
+    loadSourceSnapshot(persistenceId, progressSeqNr + 1).flatMap {
+      case None => Future.successful(0)
+      case Some(selectedSnapshot @ SelectedSnapshot(snapshotMetadata, _)) =>
+        for {
+          seqNr <- {
+            val serializedRow = serializedSnapotRow(selectedSnapshot)
+            targetSnapshotDao
+              .store(serializedRow)
+              .recoverWith { case _: R2dbcDataIntegrityViolationException =>
+                // ok, already exists
```

Comment:
Isn't this the upsert case for snapshots (so never any integrity violation?)

Reply:
right, I'll remove that